presto-verifier/src/main/java/com/facebook/presto/verifier/checksum/ColumnMatchResult.java

```diff
@@ -0,0 +1,71 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.verifier.checksum;
+
+import java.util.Objects;
+import java.util.Optional;
+
+import static com.google.common.base.MoreObjects.toStringHelper;
+import static java.util.Objects.requireNonNull;
+
+public class ColumnMatchResult
+{
+    private final boolean matched;
+    private final Optional<String> message;
+
+    public ColumnMatchResult(boolean matched, Optional<String> message)
```

Comment:
This is not the case here. Currently only floating point columns have message. For simple columns and array columns, what can be included in the message? (I don't think it would be helpful to expose the checksum that differentiate between control and test for those columns)

Reply:
Though I can add a check such that when `matched`, `message` cannot be present.

Reply:
I was about to recommend the second, but when I looked at the code:

```
return new ColumnMatchResult(
                    controlSumObject == null && testSumObject == null,
                    Optional.of(format("control(sum: %s) test(sum: %s)", controlSumObject, testSumObject)));
```

There are situations that `matched = true` and the message is not empty. I felt that extra information doesn't hurt :). We can always decide what we want to show in the UI vs not.

> (I don't think it would be helpful to expose the checksum that differentiate between control and test for those columns)

Extra information in case of failure doesn't hurt. I am fine with just printing the checksums as well in that case. My only concern is to tighten the API so that when we add new checks, it ensures that people add a message.

Reply:
Oh, now I remembered that's why I called it `message` not `errorMessage`. I'll add checksum information, but in terms of tightening the API, this is not very feasible.
- Merely adding `checkArgument` is not limiting unwanted caller behavior, and cannot be detected until at runtime.
- To tighten the API, we need to make the constructor private and to private static factory constructor such as `matched()` and `mismatched(String)`. However, since `boolean matched` can sometimes be an expression, rather than always a constant `true` or `false` at a caller, it makes those constructors very hard to use.

Reply:
> To tighten the API, we need to make the constructor private ... 

I understand this and was my first thought before I looked at the matched message use case.

> Merely adding checkArgument is not limiting unwanted caller behavior, and cannot be detected until at runtime.

I completely agree. If someone hopefully add tests or tests in prod we should catch hold of it.